src/Framework/AspNetCoreAnalyzers/src/Analyzers/RouteEmbeddedLanguage/Infrastructure/RouteStringSyntaxDetector.cs

```diff
@@ -149,18 +149,29 @@ private static bool IsAnyStringLiteral(int rawKind)
             return true;
         }
 
+        bool first = true;
         for (var node = token.Parent; node != null; node = node.Parent)
         {
-            if (HasLanguageComment(node.GetLeadingTrivia(), out identifier, out options))
+            // Deeply nested tree of strings can cause poor performance. The scenario is highly concatenated strings.
+            // GetLeadingTrivia is a hot path for this analyzer so don't call it for every parent binary node.
+            // Just check the first parent and then resume checking once out of nested binary nodes.
+            // The "ConcatString_PerformanceTest" test runs to check for this scenario.
+            if (first || (node is not BinaryExpressionSyntax && node.Parent is not BinaryExpressionSyntax))
```

Comment:
>if (first || (node is not BinaryExpressionSyntax && node.Parent is not BinaryExpressionSyntax)) [](http://example.com/codeflow?start=12&length=95)

We're skipping `GetLeadingTrivia()` here, but we're still calling `node.Parent` N^2 times when walking a set of tokens that represent a concatenation of N strings. If N is large enough, that might be an issue.

For instance, timing the following, with the source from `ConcatString_PerformanceTest`:
N = 20,000, time = ~2 secs
N = 200,000, time = ~150 secs

```csharp
var tokens = tree.GetRoot().DescendantTokens().ToArray();
foreach (var t in tree.GetRoot().DescendantTokens())
{
    RouteEmbeddedLanguage.Infrastructure.RouteStringSyntaxDetector.IsRouteStringSyntaxToken(t, model, cancellationToken: default, out _);
}
```

Reply:
If the lang comment is on a `BinaryExpressionSyntax`, are we going to ignore that comment for tokens beneath the `BinaryExpressionSyntax` node that aren't direct children?

Reply:
`node.Parent` is a plain old getter property. Its perf should be fine until you get to ludicrous scale - string concats with tens of thousands of items.